crates/turborepo-ffi/src/lockfile.rs

```diff
@@ -56,38 +64,122 @@ fn npm_transitive_closure_inner(
         ..
     } = request;
     let lockfile = NpmLockfile::load(contents.as_slice())?;
-    let dependencies = workspaces
-        .into_iter()
-        .map(|(workspace_dir, dependencies)| {
-            let closure = turborepo_lockfiles::transitive_closure(
-                &lockfile,
-                &workspace_dir,
-                dependencies.into(),
-            )?;
-            Ok((workspace_dir, proto::LockfilePackageList::from(closure)))
-        })
-        .collect::<Result<HashMap<_, _>, Error>>()?;
-    Ok(proto::WorkspaceDependencies { dependencies })
+    let dependencies = turborepo_lockfiles::all_transitive_closures(
+        &lockfile,
+        workspaces.into_iter().map(|(k, v)| (k, v.into())).collect(),
+    )?;
+    Ok(dependencies.into())
+}
+
+fn berry_transitive_closure_inner(
+    request: proto::TransitiveDepsRequest,
+) -> Result<proto::WorkspaceDependencies, Error> {
+    let proto::TransitiveDepsRequest {
+        contents,
+        workspaces,
+        resolutions,
+        ..
+    } = request;
+    let resolutions =
+        resolutions.map(|r| turborepo_lockfiles::BerryManifest::with_resolutions(r.resolutions));
+    let data = LockfileData::from_bytes(contents.as_slice())?;
+    let lockfile = BerryLockfile::new(&data, resolutions.as_ref())?;
+    let dependencies = turborepo_lockfiles::all_transitive_closures(
+        &lockfile,
+        workspaces.into_iter().map(|(k, v)| (k, v.into())).collect(),
+    )?;
+    Ok(dependencies.into())
 }
 
 #[no_mangle]
-pub extern "C" fn npm_subgraph(buf: Buffer) -> Buffer {
+pub extern "C" fn subgraph(buf: Buffer) -> Buffer {
     use proto::subgraph_response::Response;
     proto::SubgraphResponse {
-        response: Some(match npm_subgraph_inner(buf) {
+        response: Some(match subgraph_inner(buf) {
             Ok(contents) => Response::Contents(contents),
             Err(err) => Response::Error(err.to_string()),
         }),
     }
     .into()
 }
 
-fn npm_subgraph_inner(buf: Buffer) -> Result<Vec<u8>, Error> {
+fn subgraph_inner(buf: Buffer) -> Result<Vec<u8>, Error> {
     let request: proto::SubgraphRequest = buf.into_proto()?;
-    let contents = real_npm_subgraph(&request.contents, &request.workspaces, &request.packages)?;
+    let package_manager = request.package_manager();
+    let proto::SubgraphRequest {
+        contents,
+        workspaces,
+        packages,
+        resolutions,
+        ..
+    } = request;
+    let contents = match package_manager {
+        proto::PackageManager::Npm => real_npm_subgraph(&contents, &workspaces, &packages)?,
+        proto::PackageManager::Berry => turborepo_lockfiles::berry_subgraph(
```

Comment:
nit: could we use the same style of usage for both here and npm? Like either imported or with the `turborepo_lockfiles::` specified?